# **Load and Prepare HR Employee Attrition Data**

## Objectives

This notebook has been updated to reflect the **HR Employee Attrition** project data source and the required preparation steps were followed.

It main goals are to:

- load the **HR Employee Attrition dataset**
- document the **data collection** approach

- perform **basic data cleaning and validation**
- **encode categorical variables**
- **normalise key numerical features**
- save cleaned outputs for later analysis and dashboard development

## Important note on the data source

The project documentation states that the analysis uses an **HR Employee Attrition dataset** containing employee demographics, employment details, compensation data, satisfaction measures, and the target field **Attrition (Yes/No)**.


`dataset/raw/hr_employee_attrition.csv`



---

# Change working directory

   
To make relative file paths work correctly, we move from the notebook folder to the project root.


In [6]:
import os

current_dir = os.getcwd()
current_dir


'c:\\Python Projects\\Employee_attrition_analysis3\\jupyter_notebooks'

In [7]:
import os

# Navigate to project root (go up 1 level from notebook directory)
notebook_dir = os.getcwd()
project_root = os.path.dirname(notebook_dir)
os.chdir(project_root)
print(f"Working directory updated to project root: {os.getcwd()}")

Working directory updated to project root: c:\Python Projects\Employee_attrition_analysis3


In [8]:
current_dir = os.getcwd()
current_dir


'c:\\Python Projects\\Employee_attrition_analysis3'

---
# Initial Setup and Data Collection

## Data Collection

- The dataset used is the **HR Employee Attrition dataset**, selected specifically to address the business problem of understanding and reducing employee turnover.
- The dataset includes employee-level data such as demographics, employment details, compensation, satisfaction metrics, and attrition status.
- It has been sourced as an **open-source dataset** suitable for analytical and educational use.https://www.kaggle.com/code/encodedanand/employee-attrition/input
- The dataset reflects a **real-world HR analytics scenario**, where organisations aim to reduce attrition and improve workforce retention.
- No personally identifiable information (PII) is included, ensuring compliance with **ethical and GDPR considerations**.
- A known limitation is that the dataset represents a **static snapshot**, meaning trends over time cannot be analysed.

## Import libraries


In [9]:
import numpy as np
import pandas as pd


In [10]:
# Load the HR attrition dataset
# Update the filename if your CSV uses a slightly different name.

file_path = "Dataset/Raw/HR Employee Attrition.csv"
df = pd.read_csv(file_path)
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


## Quick dataset inspection

The first inspection step checks that the file loaded correctly and gives an overview of the columns, data types, and completeness of the dataset.


In [11]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

---

# Dataset Content

Based on the project brief, the dataset is expected to include fields such as:

- **Demographics**: `Age`, `Gender`
- **Employment details**: `Department`, `JobRole`, `YearsAtCompany`
- **Compensation**: `MonthlyIncome`
- **Satisfaction metrics**: `JobSatisfaction`, `WorkLifeBalance`
- **Target variable**: `Attrition`

The exact column names can vary depending on the source file, so the next step is to inspect all column names directly from the dataset.


In [12]:
df.columns.tolist()


['Age',
 'Attrition',
 'BusinessTravel',
 'DailyRate',
 'Department',
 'DistanceFromHome',
 'Education',
 'EducationField',
 'EmployeeCount',
 'EmployeeNumber',
 'EnvironmentSatisfaction',
 'Gender',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobRole',
 'JobSatisfaction',
 'MaritalStatus',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'Over18',
 'OverTime',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StandardHours',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

# Data Cleaning & Preparation

The following steps were applied to ensure data quality and consistency:

### 1. Handling missing values
- Identified missing or inconsistent values
- Applied appropriate imputation strategies (median for numerical, 'Unknown' for categorical)

### 2. Removing duplicates
- Removed redundant records to prevent bias in analysis

### 3. Encoding categorical variables
- Converted categorical variables into numerical format using one-hot encoding

### 4. Normalising key features
- Applied MinMax scaling to ensure consistent ranges across numerical features

These steps ensure the dataset is reliable, consistent, and ready for analysis and modelling.


## 1. Check missing values and duplicates


In [13]:
df.isna().sum().sort_values(ascending=False)


Age                         0
StandardHours               0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StockOptionLevel            0
MonthlyIncome               0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
MonthlyRate                 0
MaritalStatus               0
Attrition                   0
EmployeeCount               0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeNumber              0
JobSatisfaction             0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole   

In [14]:
df.duplicated().sum()
df.describe()

,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
count,1470.000000,1470.000000,1470.000000,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,...,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000
mean,36.923810,802.485714,9.192517,2.912925,1.0,1024.865306,2.721769,65.891156,2.729932,2.063946,...,2.712245,80.0,0.793878,11.279592,2.799320,2.761224,7.008163,4.229252,2.187755,4.123129
std,9.135373,403.509100,8.106864,1.024165,0.0,602.024335,1.093082,20.329428,0.711561,1.106940,...,1.081209,0.0,0.852077,7.780782,1.289271,0.706476,6.126525,3.623137,3.222430,3.568136
min,18.000000,102.000000,1.000000,1.000000,1.0,1.000000,1.000000,30.000000,1.000000,1.000000,...,1.000000,80.0,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,30.000000,465.000000,2.000000,2.000000,1.0,491.250000,2.000000,48.000000,2.000000,1.000000,...,2.000000,80.0,0.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000
50%,36.000000,802.000000,7.000000,3.000000,1.0,1020.500000,3.000000,66.000000,3.000000,2.000000,...,3.000000,80.0,1.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000
75%,43.000000,1157.000000,14.000000,4.000000,1.0,1555.750000,4.000000,83.750000,3.000000,3.000000,...,4.000000,80.0,1.000000,15.000000,3.000000,3.000000,9.000000,7.000000,3.000000,7.000000
max,60.000000,1499.000000,29.000000,5.000000,1.0,2068.000000,4.000000,100.000000,4.000000,5.000000,...,4.000000,80.0,3.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000


## 2. Handle missing values

A mixed strategy is used:

- **categorical columns**: fill missing values with `"Unknown"`
- **numerical columns**: fill missing values with the median

This keeps the dataset usable while limiting distortion from extreme values.


In [15]:
df_clean = df.copy()

categorical_cols = df_clean.select_dtypes(include=["object", "category"]).columns.tolist()
numerical_cols = df_clean.select_dtypes(include=[np.number]).columns.tolist()

for col in categorical_cols:
    df_clean[col] = df_clean[col].fillna("Unknown")

for col in numerical_cols:
    df_clean[col] = df_clean[col].fillna(df_clean[col].median())

df_clean.isna().sum().sort_values(ascending=False).head(20)


Age                         0
StandardHours               0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StockOptionLevel            0
MonthlyIncome               0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
MonthlyRate                 0
MaritalStatus               0
Attrition                   0
EmployeeCount               0
dtype: int64

this data has already been cleaned


## 3. Remove duplicate rows

Duplicate rows can distort summary statistics and model outputs, so they are removed where present.


In [16]:
df_clean = df_clean.drop_duplicates()
df_clean.shape


(1470, 35)

this means they are 1470 rows and 35 colums after cleaning

## 4. Encode categorical variables

Categorical variables are transformed into numeric format so they can be used in downstream analysis and modelling.

The notebook uses **one-hot encoding** with `drop_first=True` to reduce redundancy.


In [17]:
import pandas as pd
df_encoded = pd.get_dummies(df_clean, columns=categorical_cols, drop_first=True)
df_encoded.head()


,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,...,JobRole_Laboratory Technician,JobRole_Manager,JobRole_Manufacturing Director,JobRole_Research Director,JobRole_Research Scientist,JobRole_Sales Executive,JobRole_Sales Representative,MaritalStatus_Married,MaritalStatus_Single,OverTime_Yes
0,41,1102,1,2,1,1,2,94,3,2,...,False,False,False,False,False,True,False,False,True,True
1,49,279,8,1,1,2,3,61,2,2,...,False,False,False,False,True,False,False,True,False,False
2,37,1373,2,2,1,4,4,92,2,1,...,True,False,False,False,False,False,False,False,True,True
3,33,1392,3,4,1,5,4,56,3,1,...,False,False,False,False,True,False,False,True,False,True
4,27,591,2,1,1,7,1,40,3,1,...,True,False,False,False,False,False,False,True,False,False


## 5. Normalise key numerical features

Normalisation is applied to numerical columns so that variables measured on different scales can be compared more easily.

This is especially useful for later modelling or distance-based methods.


In [18]:
from sklearn.preprocessing import MinMaxScaler

df_prepared = df_encoded.copy()

scaler = MinMaxScaler()
available_numerical_cols = [col for col in numerical_cols if col in df_prepared.columns]

df_prepared[available_numerical_cols] = scaler.fit_transform(df_prepared[available_numerical_cols])

df_prepared[available_numerical_cols].head()


,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,0.547619,0.715820,0.000000,0.25,0.0,0.000000,0.333333,0.914286,0.666667,0.25,...,0.000000,0.0,0.000000,0.200,0.0,0.000000,0.15,0.222222,0.000000,0.294118
1,0.738095,0.126700,0.250000,0.00,0.0,0.000484,0.666667,0.442857,0.333333,0.25,...,1.000000,0.0,0.333333,0.250,0.5,0.666667,0.25,0.388889,0.066667,0.411765
2,0.452381,0.909807,0.035714,0.25,0.0,0.001451,1.000000,0.885714,0.333333,0.00,...,0.333333,0.0,0.000000,0.175,0.5,0.666667,0.00,0.000000,0.000000,0.000000
3,0.357143,0.923407,0.071429,0.75,0.0,0.001935,1.000000,0.371429,0.666667,0.00,...,0.666667,0.0,0.000000,0.200,0.5,0.666667,0.20,0.388889,0.200000,0.000000
4,0.214286,0.350036,0.035714,0.00,0.0,0.002903,0.000000,0.142857,0.666667,0.00,...,1.000000,0.0,0.333333,0.150,0.5,0.666667,0.05,0.111111,0.133333,0.117647


## 6. Validation checks

These checks confirm that the prepared dataset is suitable for the next stage of analysis.


In [19]:
print("Original shape:", df.shape)
print("Cleaned shape:", df_clean.shape)
print("Prepared shape:", df_prepared.shape)
print("\nMissing values remaining in prepared dataset:", int(df_prepared.isna().sum().sum()))


Original shape: (1470, 35)
Cleaned shape: (1470, 35)
Prepared shape: (1470, 48)

Missing values remaining in prepared dataset: 0


---

# Saving Cleaned Data

Two outputs are saved:

1. **Cleaned dataset** before encoding  
2. **Prepared dataset** after encoding and normalisation  

This supports traceability and good data management practice.


In [20]:
os.makedirs("dataset/processed", exist_ok=True)

df_clean.to_csv("dataset/processed/hr_employee_attrition_cleaned.csv", index=False)
df_prepared.to_csv("dataset/processed/hr_employee_attrition_prepared.csv", index=False)

print("Saved:")
print("- dataset/processed/hr_employee_attrition_cleaned.csv")
print("- dataset/processed/hr_employee_attrition_prepared.csv")


Saved:
- dataset/processed/hr_employee_attrition_cleaned.csv
- dataset/processed/hr_employee_attrition_prepared.csv


---

# Conclusion

This notebook has been updated to reflect the **HR Employee Attrition** data source and the required preparation steps.

Completed in this notebook:

- documented the **data collection** rationale
- aligned the notebook to the **HR attrition dataset**
- checked data structure, missing values, and duplicates
- handled missing values
- encoded categorical variables
- normalised key numerical features
- saved processed datasets for the next analysis stage

This now provides a cleaner foundation for EDA, hypothesis testing, and dashboard development.
